Fix/code-refactoring

JoaquinAkerman opened the pull request:
This branch includes a series of enhancements and refactors to the clawback feature and its associated tests.

Key changes include the addition of the clawback feature, input validation for clawback amounts, a transaction details link, and Cypress tests for the Clawback Asset page.

The branch also includes refactors to improve code readability, such as wrapping inputs in a form for validation, renaming tests for clarity, and standardizing variable declarations.

Additionally, the branch refactors the env.d.ts interface and removes unnecessary code.

Updates to the Cypress base URL, removal of unnecessary lines in tests, and enhancements to Cypress testing with the data-cy attribute are also included.

Finally, the branch includes a refactor of component attributes, the addition of the getByDataTestAttribute command to Cypress, error handling for the performClawback function, and an update to the project dependencies.

The branch concludes with a series of refactors to improve the organization of the code, including the consolidation of asset storage in localStorage, the implementation of the AssetStorageService, and the introduction of a select element for asset codes stored in localStorage, on the clawback recipe.

JoaquinAkerman committed:
Add Clawback Feature

This commit introduces the clawback feature. The page includes input fields for the asset code, issuer secret key, and clawback account, allowing users to perform a clawback operation. And a status display area to provide feedback on the operation.

JoaquinAkerman committed:
Add Input for clawback amount and checkbox for full clawback

Add an input field to specify the amount of assets to clawback. Also, add a checkbox to enable the option of clawing back all assets. When this checkbox is selected, the input field for the clawback amount is disabled, and the clawback operation will be performed for the entire balance of the specified asset in the account.

JoaquinAkerman committed:
Add validation for clawback amount exceeding balance

This commit adds a validation check in the performClawback function to ensure that the amount specified for clawback does not exceed the available asset balance in the account. If the clawback amount is greater than the available balance, an error is thrown and the transaction is not submitted.

JoaquinAkerman committed:
Add transaction details link

Added a TransactionInfo component that displays a link to the Stellar Expert Explorer for viewing transaction details upon successful clawback operation.

JoaquinAkerman committed:
Add basic Cypress tests for Clawback Asset page

Added initial Cypress tests for the Clawback Asset page. These tests check for the existence of form elements and simulate a form submission with incorrect data. More complex tests will be added.

JoaquinAkerman committed:
Add Cypress test for Clawback Asset Page

Add Cypress test that simulates the clawback operation for an account on the Clawback Asset Page. The test checks the functionality of the clawback operation, including the validation of input fields and the display of success and error messages.
Network requests to Horizon Testnet are mocked using Cypress fixtures. Which also are included on this commit

JoaquinAkerman committed:
Add visibility checks to Asset Page

Add assertions to check the visibility of the input fields before interacting with them.

JoaquinAkerman committed:
Refactor Cypress tests for Clawback Asset Page and add new test

This commit refactors the Cypress tests for the Clawback Asset Page to use variables instead of repeated string literals.
In addition, a new test has been added to check full clawback operation

JoaquinAkerman committed:
Add 'required' attribute to Input component

Now, with the 'required' attribute, the Input component prevents empty field submissions, reducing unnecessary requests.

JoaquinAkerman committed:
Wrap Inputs in Form and Add Required Attribute

In this commit, we have wrapped the Input components inside a form element. This allows us to use the 'required' attribute on the inputs, which prevents the form from being submitted when these inputs are empty.

JoaquinAkerman committed:
Add Cypress Test Case for Transaction Failure

In this commit, we have added a new test case in Cypress to verify that an error message is displayed when a transaction fails. We have also added a fixture for a failed transaction. This fixture is used to mock the network request for the transaction and return a simulated error response.

JoaquinAkerman committed:
Remove unnecessary statement and import

In this commit, an unnecessary @html statement and an unused import of 'Transaction' from 'stellar-sdk' were removed from the Svelte component.

The @html statement was not contributing to the functionality, and the 'Transaction' import was not being used anywhere in the code.

This do not affect the functionality of the component.

JoaquinAkerman committed:
Refactor Cypress tests and rename for clarity

In this commit, the Cypress test was refactored to use the baseUrl from the Cypress configuration instead of hardcoding the URL.

JoaquinAkerman committed:
Refactor Cypress test to use more stable selectors

The Cypress test was refactored to use more stable selectors. Previously, the test was relying on a class selector.

The selector was replaced with an assertion  based on the content of the element.

JoaquinAkerman committed:
Rename test for better clarity

The new name provides a more accurate description of the test's functionality, making it easier to understand what is being tested and why.

JoaquinAkerman committed:
Add TypeScript support and standardize variable declarations

In this commit, TypeScript support was added to several Svelte components by adding the `lang="ts"` attribute to the `script` tags. Additionally, variable declarations were standardized across these components.

JoaquinAkerman committed:
Refactor env.d.ts interface

JoaquinAkerman committed:
Remove unnecessary @html from status display

JoaquinAkerman committed:
Refactor in +page.svelte of clawback

In this commit, TypeScript support was added to the +page.svelte file of the clawback feature by adding the `lang="ts"` attribute to the `script` tag. Additionally, several variable declarations were refactored to include TypeScript type annotations. An id was also added to the Card component for better element targeting.

JoaquinAkerman committed:
Wrap inputs in form for validation

Wraps all input fields inside a form element. This allows us to leverage the built-in HTML form validation by using the 'required' attribute on inputs. This prevents the form from being submitted when the required inputs are empty.

JoaquinAkerman committed:
Remove Cypress test for empty input submission

This commit removes the Cypress test that was checking for form submission with empty inputs. With the recent changes to wrap inputs inside a form and use the 'required' attribute, the form now prevents submission when required inputs are empty.

JoaquinAkerman committed:
Merge branch 'main' into code-refactoring

JoaquinAkerman committed:
Update Cypress base URL

This commit updates the base URL in the Cypress configuration to include the '/stellar-chef' path. This change ensures that Cypress tests start from the correct path when running end-to-end tests.

JoaquinAkerman committed:
Remove unnecessary lines in clawbackTests.cy.ts

eloptimistadelcode commented on cypress/e2e/asset-issuanceTest.cy.ts:
```diff
@@ -1,10 +1,9 @@
 
 const visitAssetIssuancePage = (): void => {
-  cy.visit(`${BASE_URL}/recipe/asset-issuance`);
+  cy.visit(Cypress.config().baseUrl + '/recipe/asset-issuance');
```
```suggestion
  cy.visit('/recipe/asset-issuance');
```

JoaquinAkerman committed:
Rename test files

This commit addresses redundancy in file naming by removing the unnecessary "Test" prefix from asset-issuanceTest.cy.ts and clawbackTest.cy.ts

JoaquinAkerman committed:
Declare $env/static/public module and update config

This commit includes the declaration of the $env/static/public module in the svelte-env.d.ts file and updates the TypeScript and ESLint configurations to recognize this module.

JoaquinAkerman committed:
Update Stellar network URL and passphrase imports

This commit updates the way the Stellar network URL and passphrase are imported in utils.ts. Instead of using import.meta.env, the values are now imported from the $env/static/public module. This change aligns with the recent configuration updates.

JoaquinAkerman committed:
Enhance Cypress testing with data-cy attribute

"This commit includes changes across four component files where unnecessary id attributes were removed and data-cy attributes were added to improve Cypress testing practices."

JoaquinAkerman committed:
Refactor component attributes and add data-cy

"Updated component attributes for better clarity. Added data-cy attributes to Cypress testing.

JoaquinAkerman committed:
Add getByDataTestAttribute command to Cypress

Add a new Cypress command, getByDataTestAttribute, to simplify the selection of elements by their data-cy attribute.

JoaquinAkerman committed:
Components refactor

Move data-cy attribute to anchor element for improved testing practices and rename data-cy property for increased clarity

JoaquinAkerman committed:
Update tests

Update tests for assertions based on data-cy on anchor elements

JoaquinAkerman committed:
Add error handling to performClawback function

This commit introduces error handling to the performClawback function. Now, if the assetCode, issuer secret key, or the clawback account is invalid, an error message will be displayed. This commit also adds data-cy attributes to the form inputs for testing purposes.

JoaquinAkerman committed:
Add data-cy types to custom command

This commit add specific types for the data-cy attribute in the custom Cypress command 'getByDataTestAttribute'. This will help ensure that only valid data-cy attributes are used when calling the command.

JoaquinAkerman committed:
Update data-cy typings

JoaquinAkerman committed:
Update clawback.cy.ts to use custom command

This commit updates the clawback.cy.ts test file to use the custom Cypress command 'getByDataTestAttribute'. This change simplifies the test code and makes it more robust by allowing us to select elements based on their 'data-cy' attribute.

JoaquinAkerman committed:
Remove svelte-env.d.ts file

JoaquinAkerman committed:
Update environment variables for public Stellar network

JoaquinAkerman committed:
Remove unnecessary include from tsconfig.json

JoaquinAkerman committed:
Update project dependencies

JoaquinAkerman committed:
Add Cypress test fixtures and update asset creation test

Updated all asset issuance requests to be mocked, eliminating reliance on the testnet. Included fixtures for the mocked requests. Also eliminated timeouts on tests.

JoaquinAkerman committed:
Merge branch 'main' into code-refactoring

JoaquinAkerman committed:
Remove unnecessary comments from code

JoaquinAkerman committed:
Rename fixture files for consistency

Renamed the fixture file from 'horizon-testnet-stellar-friendbot-Response' to 'horizonTestnetStellarFriendbotResponse.json' to maintain consistent naming conventions across the project.

JoaquinAkerman committed:
Consolidate asset storage in localStorage

Refactored the Asset and LocalStorage classes to consolidate asset storage in localStorage.
Previously, each asset was stored under an individual key in localStorage. Now, all assets are stored under a single key named 'assets'. The value is a stringified array of the IAsset objects, allowing multiple assets to be stored together.

JoaquinAkerman committed:
Implement AssetStorageService

Implemented the AssetStorageService in +page.svelte to handle asset storage.

The created asset and the issuer keys are now saved using the AssetStorageService. This allows for efficient storage and retrieval of the created asset information

JoaquinAkerman committed:
Factor the Asset and LocalStorage classes

Updated the set method in the Asset class to utilize the existing instance of LocalStorage and refined the type definition for the value parameter in the set method of the LocalStorage class to accurately reflect the expected input of an array of objects or an empty array at initialization.

JoaquinAkerman committed:
Change visibility of method in Asset class

Updated the visibility of the get method in the Asset class from private to public. This change allows the get method to be accessed from outside the class, specifically for use in the autocomplete functionality of the clawback feature.

JoaquinAkerman committed:
Add autocomplete for assetCode and issuerSecretKey

Updated the asset code input field to support autocomplete from existing assets stored in localStorage.  Additionally, a reactive statement was added to update the issuerSecretKey whenever an assetCode is selected from the autocomplete.

JoaquinAkerman committed:
Remove unnecessary null check in LocalStorage.ts

JoaquinAkerman committed:
Remove repeated code in Asset class

JoaquinAkerman committed:
Replace auto-complete input with select for asset code

Replaced the auto-complete input for asset code with a select element. The select options are populated with all the assets saved in localStorage.

JoaquinAkerman committed:
Refactor Cypress command type declaration

Moved the type declaration for the `getByDataTestAttribute` Cypress command to a `index.d.ts` file.
Removed the typing for the `getByDataTestAttribute` custom Cypress command.

JoaquinAkerman committed:
Remove unnecessary code in Asset.ts
